### PR TITLE
Convert PreActivation to functional style

### PR DESCRIPTION
**Pull request by jasonaden**
This adds predicability and makes the change for coalescing of guards easier to implement. Also simplifies the testing interface.

### CODE REVIEW

**Comment by jasonaden** [comment]
[Presubmit](http://test/OCL:215599738:BASE:215599790:1538592704032:ddf3aa3f)

**Review by benlesh — APPROVED** [review]
LGTM with a few nits

File: packages/router/src/operators/check_guards.ts [hunk 1]
```diff
@@ -6,19 +6,154 @@
...
+
+function runCanActivateChild(
+    futureRSS: RouterStateSnapshot, path: ActivatedRouteSnapshot[],
+    moduleInjector: Injector): Observable<boolean> {
+  const futureARS = path[path.length - 1];
+
+  const canActivateChildGuards = path.slice(0, path.length - 1)
+                                     .reverse()
+                                     .map(p => getCanActivateChild(p))
+                                     .filter(_ => _ !== null);
```
[the remark this answers is not in the document]
Comment (jasonaden) [reply packages/router/src/operators/check_guards.ts]: I'll leave this one as it's a small array and executes infrequently, so I think the extra bytes aren't worth it for this one.

File: packages/router/src/operators/check_guards.ts [hunk 2]
```diff
@@ -6,19 +6,154 @@
...
+   * This should fire off `ActivationStart` events for each route being activated at this
+   * level.
+   * In other words, if you're activating `a` and `b` below, `path` will contain the
+   * `ActivatedRouteSnapshot`s for both and we will fire `ActivationStart` for both. Always
+   * return
+   * `true` so checks continue to run.
+   */
+function fireActivationStart(
+    snapshot: ActivatedRouteSnapshot | null,
+    forwardEvent?: (evt: Event) => void): Observable<boolean> {
```
[the remark this answers is not in the document]
Comment (jasonaden) [reply packages/router/src/operators/check_guards.ts]: 👍 Let's keep an eye on this with the upcoming PRs. We might be able to refactor this along with some of the work happening now.

**Pull request closed without merging**